### ISSUE TITLE
ipfs add doesnt handle broken symlinks

### ISSUE THREAD

**Issue by whyrusleeping** [ipfs/go-ipfs#979]
running `ipfs add` on a directory containing broken symlinks will cause unexpected behaviour. Depending on a few different factors, this may cause a hang or just an early exit and leave the operation unfinished.

**Comment by jbenet** [comment]
We should start with a test case



—

**Comment by cryptix** [comment]
@whyrusleeping can you share some reproduction insight on this one? I tried the following but it works as expected (for me at least).


```
[cryptix@higgs /tmp/ipfsSymlink] ls
[cryptix@higgs /tmp/ipfsSymlink] dd if=/dev/urandom of=test bs=1k count=100
100+0 records in
100+0 records out
102400 bytes (102 kB) copied, 0.00661302 s, 15.5 MB/s
[cryptix@higgs /tmp/ipfsSymlink] ln -s test simpleLink
[cryptix@higgs /tmp/ipfsSymlink] ipfs add *
added QmZePokqaRebXVmcWiSJyZfiqFqUSoHggWErbE9Q2H1Vex simpleLink
added QmZePokqaRebXVmcWiSJyZfiqFqUSoHggWErbE9Q2H1Vex test
[cryptix@higgs /tmp/ipfsSymlink] ln -s please/please/dont/exist broken_link && ipfs add broken_link
Error: open broken_link: no such file or directory

Use 'ipfs add --help' for information about this command
[cryptix@higgs /tmp/ipfsSymlink] cd ..
[cryptix@higgs /tmp] ipfs add -r ipfsSymlink
Error: Post http://127.0.0.1:5001/api/v0/add\?encoding\=json\&progress\=true\&r\=true\&stream-channels\=true: open ipfsSymlink/broken_link: no such file or directory
```

**Comment by whyrusleeping** [comment]
repro steps:

```
$ mkdir a
$ echo "some real content" > a/b
$ ln -s /does/not/exist a/badlink
$ ipfs add -r a
```

This example should fail with an error, but given a larger directory (an entire docker filesystem, with a bad link somewhere) generally causes it to hang

**Comment by wking** [comment]
On Sun, Mar 29, 2015 at 12:10:55PM -0700, Jeromy Johnson wrote:
> running `ipfs add` on a directory containing broken symlinks will
> cause unexpected behaviour.

With #1413 I'm closing in on adding symlinks to IPFS as symlinks
(instead of dereferencing them).  I think that means we can close this
issue in favor of #616.

**Comment by jbenet** [comment]
@wking people often search with open:issues in the filters, or subscribe. no harm in waiting to close till it's merged

**Comment by whyrusleeping** [comment]
closed by #1627
